src/solc_0.8/assetERC1155/AssetBaseERC1155.sol

```diff
@@ -0,0 +1,668 @@
+                _packedTokenBalance[owner][bin] = _packedTokenBalance[owner][bin].updateTokenBalance(
+                    index + i,
+                    supplies[j],
+                    ObjectLib32.Operations.ADD
+                );
+            }
+        }
+    }
+
+    function _transferFrom(
+        address from,
+        address to,
+        uint256 id,
+        uint256 value
+    ) internal returns (bool) {
+        address sender = _msgSender();
+        bool authorized = from == sender || isApprovedForAll(from, sender);
+
+        require(authorized, "OPERATOR_!AUTH");
+        if (value > 0) {
+            (uint256 bin, uint256 index) = id.getTokenBinIndex();
+            _packedTokenBalance[from][bin] = _packedTokenBalance[from][bin].updateTokenBalance(
+                index,
+                value,
+                ObjectLib32.Operations.SUB
+            );
+            _packedTokenBalance[to][bin] = _packedTokenBalance[to][bin].updateTokenBalance(
+                index,
+                value,
+                ObjectLib32.Operations.ADD
+            );
+        }
+
+        emit TransferSingle(isTrustedForwarder(msg.sender) ? from : sender, from, to, id, value);
+        return true;
+    }
+
+    function _mint(
+        bytes32 hash,
+        uint256 supply,
+        uint8 rarity,
+        address operator,
+        address owner,
+        uint256 id,
+        bytes memory data,
+        bool extraction
+    ) internal {
+        uint256 uriId = id & ERC1155ERC721Helper.URI_ID;
+        if (!extraction) {
+            require(uint256(_metadataHash[uriId]) == 0, "ID_TAKEN");
+            _metadataHash[uriId] = hash;
+            require(rarity < 4, "RARITY>=4");
+            bytes memory pack = new bytes(1);
+            pack[0] = bytes1(rarity * 64);
+            _rarityPacks[uriId] = pack;
+        }
+        _mint(operator, owner, id, supply, data);
+    }
+
+    function _mint(
+        address operator,
+        address account,
+        uint256 id,
+        uint256 amount,
+        bytes memory data
+    ) internal {
+        (uint256 bin, uint256 index) = id.getTokenBinIndex();
+        _packedTokenBalance[account][bin] = _packedTokenBalance[account][bin].updateTokenBalance(
+            index,
+            amount,
+            ObjectLib32.Operations.REPLACE
+        );
+
+        emit TransferSingle(operator, address(0), account, id, amount);
+        require(_checkOnERC1155Received(operator, address(0), account, id, amount, data), "TRANSFER_REJECTED");
+    }
+
+    function _extractERC721From(
+        address sender,
+        uint256 id,
+        address to
+    ) internal returns (uint256 newId) {
+        require(to != address(0), "TO==0");
+        require(balanceOf(to, id) == 1, "!NFT");
+        uint32 tokenCollectionIndex = _nextCollectionIndex[id];
+        bytes32 metaData = _metadataHash[id & ERC1155ERC721Helper.URI_ID];
+        _nextCollectionIndex[id] = tokenCollectionIndex + 1;
```

Comment:
this variables _nextCollectionIndex are not bridged so if we extract an asset on L1, we won't have this on L2

Reply:
`extractERC721From` is defined in `AssetUpgrader` which is only on L1